### PR TITLE
feat: gto assign / deprecate / register  —auto-push

### PR DESCRIPTION
**Pull request by francesco086**
- close https://github.com/iterative/gto/issues/282
- close https://github.com/iterative/gto/issues/283

### CODE REVIEW

**Comment by francesco086** [comment]
@aguschin if you could please give a feedback on this...?

If it is ok, then I would apply the same (add the `--auto-push` option) to all writing commands as part of this PR.

File: gto/git_utils.py
Comment (aguschin) [inline gto/git_utils.py:76]: I think it would be good if we would be verbose here and communicate how the command changed the world outside of it :)

Like always printing this to stdout:
```shell
$ gto register model --auto-push
Created Git tag `model@v0.0.1`
Running `git push origin model@v0.0.1`
Successfully pushed Git tag(s) to the remote.
```

Or even just printing the real command output (I'm sure `GitPython` calls `git` CLI underneath, so maybe we can just get that output and print it). In this case it would behave just as if you run that command, which is well recognizeable and familiar.

**Review by aguschin — APPROVED** [review]
LGTM! Like your mocking tests!

File: gto/git_utils.py
Comment (francesco086) [reply gto/git_utils.py:83]: that's what I would guess too, why?

**Comment by francesco086** [comment]
While I wait for your feedback regarding the echoed message, I extend the same option (`--auto-push`) to all the other writing commands

File: gto/git_utils.py
Comment (aguschin) [reply gto/git_utils.py:83]: Sorry, didn't get the question :)
I assume this would fail cause you invoke `push`, but `remote_url` is None, so it don't know where to push to. Am I wrong?

File: gto/registry.py
Comment (aguschin) [inline gto/registry.py:568]:
```suggestion
                echo(f"Running `git push origin {tag_name}`")
```

File: gto/git_utils.py
Comment (aguschin) [reply gto/git_utils.py:76]: Yes, looks good!

File: gto/git_utils.py
Comment (francesco086) [reply gto/git_utils.py:83]: `remote_url` is just used in the logging (just a safe wrapper around `remote.url`), I don't think it has nothing to do with failure of `remote.push`.

But I think you are onto something...?

Comment (aguschin) [reply gto/git_utils.py:83]: ok, I was trying to check if there are cases when you actually get a remote without URL (`remote_url == "unknown"` eventually), in this case `remote.push` could fail (because there is no actual remote to use). In this case it would be good to throw some GTO error (like `WrongArg` for example), and not a GitPython one.

**Comment by francesco086** [comment]
I covered all cli writing commands except for:
- `annotate`
- `remove`

File: tests/test_api.py
Comment (francesco086) [inline tests/test_api.py:393]: Before merging: could you please help me here?

Comment (francesco086) [reply tests/test_api.py:393]: forgot to cite @aguschin

File: gto/registry.py
Comment (aguschin) [inline gto/registry.py:491]: You have these prints across the code. Is this the debug output you forgot to remove?

Comment (francesco086) [reply gto/registry.py]: yes :/ I should really find out why the debugger does not work. I removed them now, thanks

File: tests/test_api.py
Comment (aguschin) [reply tests/test_api.py:393]: Next lines are
```python
        gto.api.unassign(
            repo.working_dir, name="model", stage="dev", ref="HEAD", auto_push=True
        )
```
You unassign a `stage` from a `model version`. Generally, you need to specify the version OR `ref`, that pins down the right version. With `Envs-like` approach we use now, you could remove this limitation, since there is only one model version in a stage at particular moment. But, GTO also tries to support `Kanban-like` approach and `Label-like` approach - [docs](https://github.com/iterative/gto#enabling-multiple-versions-in-the-same-stage-workflow). In those cases you need to keep this limitation in place. That's why it's here. Also, for the reference, we have an [issue somehow related to this](https://github.com/iterative/gto/issues/272).

Comment (francesco086) [reply tests/test_api.py]: thanks for the clarification, I removed the comment from the code

**Comment by francesco086** [comment]
@aguschin I think it's ready to merge?

File: gto/cli.py
Comment (aguschin) [inline gto/cli.py:569]: Looks like `assign --auto-push` doesn't push the first Git tag that is created to register a version:
```bash
$ gto promote ohmypromotion --stage dev --auto-push
Created git tag 'ohmypromotion@v0.0.1' that registers version
To push the changes upstream, run:
    git push origin ohmypromotion@v0.0.1
Created git tag 'ohmypromotion#dev#1' that assigns stage to version 'v0.0.1'
Running `git push origin ohmypromotion#dev#1`
Successfully pushed git tag ohmypromotion#dev#1 to the remote.
```

Comment (aguschin) [reply gto/cli.py:569]: Open question: is it better to do it like this (create tag, push tag, create tag, push tag), or do it together (create both tags, push both)?

File: gto/cli.py
Comment (aguschin) [inline gto/cli.py:630]: A bug 🤔 
```shell
$ gto register bug --auto-push
Created git tag 'bug@v0.0.1' that registers version
Running `git push origin bug@v0.0.1`
Successfully pushed git tag bug@v0.0.1 to the remote.

$ gto deprecate bug --auto-push --delete
Deleted git tag 'bug@v0.0.1'
Running `git push origin bug@v0.0.1`
❌ Unexpected error: 'str' object has no attribute 'new_message_handler'
Please report it here running with '--traceback' flag: <https://github.com/iterative/gto/issues>
```

File: gto/git_utils.py
Comment (aguschin) [inline gto/git_utils.py:71]: Something to consider: it could be that we have some Git tags locally, but don't have them in remote. Then if you run `--auto-push` in locally cloned repo, you'll decide what to do based on the local repo, but you will try to apply it to the remote one. It could be an error in some cases (you try to remove a remote Git tag that doesn't exist in remote). It could be a incorrect decision (you deprecate a version that doesn't exist in the remote). Looks too complicated to resolve at Friday evening 🍻

**Comment by aguschin** [comment]
Overall looks good, but found couple of bugs @francesco086!

File: gto/cli.py
Comment (francesco086) [reply gto/cli.py:630]: argh! well spotted... My expectations for the test was wrong! I think I fixed it with this commit https://github.com/iterative/gto/pull/281/commits/ea2df8bfb63d7311b6d481cddb560fe23e7acbd3

File: gto/git_utils.py
Comment (francesco086) [reply gto/git_utils.py:71]: "From great power comes great responsibility" I'd say :D
In the end the auto-push is something you can choose to use or not at every call, as you can choose or not to git push a tag. But if this were not (very) useful, why would you echo a suggestion to push the created tags?

In practice I find hard to imagine to work with a local repo out of sync with the remote. This is the CI/CD practice, isn't it?
In fact I would suggest to be able to set auto-push always true in a config file...

File: gto/cli.py
Comment (aguschin) [reply gto/cli.py:569]: For some reason it doesn't fail if `--stage` arg is not provided, causing this:
```shell
$ gto assign bugfixed --auto-push
Created git tag 'bugfixed@v0.0.1' that registers version
Running `git push origin bugfixed@v0.0.1`
Successfully pushed git tag bugfixed@v0.0.1 to the remote.
Created git tag 'bugfixed#None#1' that assigns stage to version 'v0.0.1'
Running `git push origin bugfixed#None#1`
Successfully pushed git tag bugfixed#None#1 to the remote.
❌ Cannot parse tag name 'bugfixed#None#1'
```

File: gto/git_utils.py
Comment (aguschin) [reply gto/git_utils.py:71]: > In practice I find hard to imagine to work with a local repo out of sync with the remote.

This could happen by chance. But ok, I guess we'll need to accommodate for this in the future. This may happen as part of implementing `gto status` command, so I'll write it [in that ticket](https://github.com/iterative/gto/issues/115#issue-1201663040).

> In fact I would suggest to be able to set auto-push always true in a config file...

Good idea!

File: gto/cli.py
Comment (francesco086) [reply gto/cli.py:569]: mmm... I can't see how it can be related to what I have done in this PR. Behavior is same without the `--auto-push`, can you confirm? This is something that probably should be fixed in a separate PR?

Perhaps now is more urgent because one may push invalid tags (like `bugfixed#None#1`)?

**Comment by francesco086** [comment]
> Overall looks good, but found couple of bugs @francesco086!

Thanks a lot @aguschin , I think I solved them :) except [this](https://github.com/iterative/gto/pull/281#discussion_r990294060), but I don't it's related to this PR - please correct me if I am wrong

File: gto/cli.py
Comment (aguschin) [reply gto/cli.py:569]: Yes, it doesn't depend on `--auto-push`. We can do it in a separate PR for sure. Created a separate issue for that https://github.com/iterative/gto/issues/285

File: gto/git_utils.py
Comment (aguschin) [reply gto/git_utils.py:71]: Example of the currently misleading behavior. Consider you have a git tag `bug@v0.0.1` in remote, but don't have it locally. Then:

```
$ gto register bug --auto-push
Created git tag 'bug@v0.0.1' that registers version
Running `git push origin bug@v0.0.1`
WARNING:git.remote:Error lines received while fetching: error: failed to push some refs to 'https://github.com/aguschin/example-gto'
Successfully pushed git tag bug@v0.0.1 on remote.
```

I think we should fail at push, printing something like: `The Git tag bug@v0.0.1 already exists on remote.` WDYT?

File: gto/cli.py
Comment (aguschin) [inline gto/cli.py:521]: Let's mark it with `experimental` for now? I anticipate some problems when remote diverges from local, so would be good to warn users we're testing and developing this functionality.
```suggestion
        False, "--auto-push", is_flag=True, help="Push created tag automatically (experimental)"
```

Comment (francesco086) [reply gto/cli.py]: good idea! I implemented the change in this https://github.com/iterative/gto/pull/281/commits/a2b0b30c0ef24dedd5cf85350186100faa3f8a65 (the change had to be done in 3 different places)

File: gto/git_utils.py
Comment (francesco086) [reply gto/git_utils.py:71]: I agree that it would be more user friendly, therefore desirable. But on the other hand, the same exact error would happen if the user does the git push manually, right? So, as the user can choose if to do git push or not, can also choose whether to add the option `--auto-push` or not.

In conclusion, I don't see much harm in it, but one could make error messages easier to interpret, I agree.

Comment (francesco086) [reply gto/git_utils.py:71]: Shall we create a new issue for this?

Comment (aguschin) [reply gto/git_utils.py:71]: Ok, think I got your point. As I see the difference
1. We're communicating it incorrectly. We say `Successfully pushed git tag...`. E.g. the user may assume the CI is triggered in the remote. But the tag was not pushed, and CI was not triggered. So we're confusing user with this message.
2. Often, a user registers a version or assigns a stage to trigger CI. In this case the goal wasn't achieved, so we should exit with code 1. It could be that user don't want any CI and just want the remote repo to be in this state (so as the Git tag was pushed, we don't need to do anything and it is fine). In this case it may be ok, but I think triggering the CI is more frequent goal here. We may add some tag (`--no-fail-if-tags-exist-in-remote` e.g.), but I think it's overkill now.

Comment (francesco086) [reply gto/git_utils.py:71]: ok, then is it ok if I apply the following change: if `git push` fails, then echo an appropriate message and exit with code 1?

Comment (francesco086) [reply gto/git_utils.py:71]: appropriate message being something like "Execution of `git push origin bug@v0.0.1` failed. Make sure your local repository is in sync with the remote."

ok?

Comment (aguschin) [reply gto/git_utils.py:71]: Yes, totally. Thanks 🙏🏻

Comment (francesco086) [reply gto/git_utils.py:71]: This should be it: https://github.com/iterative/gto/pull/281/commits/0f39722686ab42f29a3d60de37f904089e013975

**Comment by francesco086** [comment]
Please wait before merging (in case everything is fine), I would like to clean up a little bit the tests

**Comment by francesco086** [comment]
> Please wait before merging (in case everything is fine), I would like to clean up a little bit the tests

finished! https://github.com/iterative/gto/pull/281/commits/3858a3f7588b7fd92e5e3c4d3ca7753029b8c96d

**Comment by aguschin** [comment]
Great! It took some time to fix different moments, but I'm 100% sure it was worth it) Thanks a lot for the implementation!

**Pull request merged**
